apollo-router/src/configuration/cors.rs

```diff
@@ -84,91 +271,371 @@ impl Cors {
+        CorsService {
+            inner: service,
+            config: self.config.clone(),
+        }
+    }
+}
+
+/// Our custom CORS service that handles per-origin configuration
+#[derive(Clone, Debug)]
+pub(crate) struct CorsService<S> {
+    inner: S,
+    config: Cors,
+}
+
+impl<S, ReqBody, ResBody> Service<Request<ReqBody>> for CorsService<S>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>> + Send + 'static,
+    S::Future: Send + 'static,
+    S::Error: Send + 'static,
+    ReqBody: Send + 'static,
+    ResBody: Send + 'static + Default,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        let request_origin = req.headers().get(ORIGIN).cloned();
+        let is_preflight = req.method() == http::Method::OPTIONS;
+        let config = self.config.clone();
+        let request_headers = req.headers().get(ACCESS_CONTROL_REQUEST_HEADERS).cloned();
+
+        // Intercept OPTIONS requests and return preflight response directly
+        if is_preflight {
+            let mut response = Response::builder()
+                .status(http::StatusCode::OK)
+                .body(ResBody::default())
+                .unwrap();
+            // Find matching origin configuration
+            let policy = Self::find_matching_policy(&config, &request_origin);
+            // Add CORS headers for preflight
+            Self::add_cors_headers(
+                &mut response,
+                &config,
+                &policy,
+                &request_origin,
+                true,
+                request_headers,
+            );
+            return Box::pin(async move { Ok(response) });
+        }
+
+        let fut = self.inner.call(req);
+        Box::pin(async move {
+            let mut response = fut.await?;
+            // Find matching origin configuration
+            let policy = Self::find_matching_policy(&config, &request_origin);
+            // Add CORS headers for non-preflight
+            Self::add_cors_headers(
+                &mut response,
+                &config,
+                &policy,
+                &request_origin,
+                false,
+                request_headers,
+            );
+            Ok(response)
+        })
+    }
+}
+
+impl<S> CorsService<S> {
+    /// Find the matching policy for a given origin
+    fn find_matching_policy<'a>(
+        config: &'a Cors,
+        origin: &'a Option<http::HeaderValue>,
+    ) -> Option<&'a Policy> {
+        let origin_str = origin.as_ref()?.to_str().ok()?;
+
+        // Check for exact origin matches first
+        if let Some(policies) = &config.policies {
+            for policy in policies.iter() {
```

Comment:
Do we also need to check `allow_any_origin` here?

Reply:
(Probably not: https://github.com/apollographql/router/pull/7853/files#r2215958170)